websocket/jetty12-ee10/src/main/java/jenkins/websocket/Jetty12EE10Provider.java

```diff
@@ -0,0 +1,180 @@
+            return null;
+        }
+        return new Handler() {
+            @Override
+            public Future<Void> sendBinary(ByteBuffer data) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendBinary(data, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendBinary(ByteBuffer partialByte, boolean isLast) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendPartialBinary(partialByte, isLast, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendText(String text) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendText(text, f);
+                return f;
+            }
+
+            @Override
+            public Future<Void> sendPing(ByteBuffer applicationData) {
+                Callback.Completable f = new Callback.Completable();
+                session().sendPing(applicationData, f);
+                return f;
+            }
+
+            @Override
+            public void close() {
+                session().close();
+            }
+
+            private Session session() {
+                Session session = (Session) listener.getProviderSession();
+                if (session == null) {
+                    throw new IllegalStateException("missing session");
+                }
+                return session;
+            }
+        };
+    }
+
+    private static Object createWebSocket(JettyServerUpgradeRequest req, JettyServerUpgradeResponse resp) {
+        Listener listener = (Listener) req.getHttpServletRequest().getAttribute(ATTR_LISTENER);
+        if (listener == null) {
+            throw new IllegalStateException("missing listener attribute");
+        }
+        return new SessionListener(listener);
+    }
+
+    public static class SessionListener implements Session.Listener.AutoDemanding {
+        private final Listener listener;
+
+        public SessionListener(Listener listener) {
+            this.listener = Objects.requireNonNull(listener);
+        }
+
+        @Override
+        public void onWebSocketOpen(Session session) {
+            listener.onWebSocketConnect(session);
+        }
+
+        @Override
+        public void onWebSocketBinary(ByteBuffer payload, Callback callback) {
+            try {
+                int len = payload.remaining();
+                if (payload.hasArray()) {
+                    listener.onWebSocketBinary(payload.array(), payload.arrayOffset() + payload.position(), len);
+                    payload.position(payload.position() + len);
+                } else {
+                    byte[] bytes = new byte[len];
+                    payload.get(bytes);
+                    listener.onWebSocketBinary(bytes, 0, len);
+                }
```

Comment:
Likely the most important part of this PR from a reviewers' perspective. Note that eventually, we can clean up this code and make the Jenkins interface take a raw `ByteBuffer`, which will be cleaner. But since that involves making breaking API changes, I chose not to tackle that as part of this PR to minimize risk and disruption. Once we are on EE 10, I can come back to clean this up, which will be a minor breaking change in and of itself.